File: Heap/BinaryHeap.cpp
// https://www.geeksforgeeks.org/binary-heap/
#include <iostream>
#include <climits>

void swap(int&, int&);

class MinHeap {
private:
    int *harr;
    int capacity;
    int heap_size;
public:
    // constuctor
    MinHeap(int capacity);

    void MinHeapify(int);

    int parent(int i) { return (i-1) / 2; }

    int left(int i) { return 2*i+1; }

    int right(int i) { return 2*i+2; }

    int extractMin();

    void decreaseKey(int, int);

    int getMin() { return harr[0]; }

    void deleteKey(int);

    void insertKey(int);

    void print_heap();
};

MinHeap::MinHeap(int cap) {
    heap_size = 0;
    capacity = cap;
    harr = new int[cap];
}

void MinHeap::insertKey(int k) {
    if (heap_size == capacity)
        return;

    int i = heap_size;
    heap_size++;
    harr[i] = k;

    while (i != 0 && harr[parent(i)] > harr[i]) {
        swap(harr[i], harr[parent(i)]);
        i = parent(i);
    }
}

void MinHeap::decreaseKey(int i, int new_val) {
    harr[i] = new_val;
    while (i != 0 && harr[parent(i)] > harr[i]) {
        swap(harr[i], harr[parent(i)]);
        i = parent(i);
    }
}

int MinHeap::extractMin() {
    if (heap_size <= 0)
        return INT_MAX;
    if (heap_size == 1) {
        heap_size--;
        return harr[0];
    }
    int root = harr[0];
    harr[0] = harr[heap_size-1];
    heap_size--;
    MinHeapify(0);
    return root;
}

void MinHeap::deleteKey(int i) {
    decreaseKey(i, INT_MIN);
    extractMin();
}

void MinHeap::MinHeapify(int i) {
    int l = left(i);
    int r = right(i);
    int smallest = i;
    if (l < heap_size && harr[l] < harr[i])
        smallest = l;
    if (r < heap_size && harr[r] < harr[smallest])
        smallest = r;
    if (smallest != i) {
        swap(harr[i], harr[smallest]);
        MinHeapify(smallest);
    }
}

void MinHeap::print_heap() {
    std::cout << " >> ";
    for (int i = 0; i < heap_size; i++)
        std::cout << harr[i] << " ";
    std::cout << heap_size << '\n';
}

void swap(int& x, int& y) {
    int temp = x;
    x = y;
    y = temp;
}

int main()
{
    MinHeap h(11);
    h.insertKey(3);
    h.insertKey(2); // 2, 3
    h.deleteKey(1);
    h.insertKey(15);
    h.insertKey(5);
    h.insertKey(4);
    h.insertKey(45);
    h.print_heap();
    std::cout << h.extractMin() << " ";
    h.print_heap();
    std::cout << h.getMin() << " ";
    h.print_heap();
    h.decreaseKey(2, 1);
    h.print_heap();
    std::cout << h.getMin();
    return 0;
}

File: Heap/BinaryHeap.py
from heapq import heappush, heappop

class MinHeap:
    """Implement heap by list"""
    def __init__(self):
        self.h = []
        self.size = 0

    @staticmethod
    def parent(i):
        return (i - 1) // 2

    @staticmethod
    def left(i):
        return 2 * i + 1

    @staticmethod
    def right(i):
        return 2 * i + 2

    def extract_min(self):
        if self.size <= 0:
            return None
        if self.size == 1:
            self.size -= 1
            return self.h.pop()
        root = self.h[0]
        self.size -= 1
        self.h[0] = self.h.pop() # [-1]
        self.min_heapify(0)
        return root

    def decrease_val(self, i, val):
        # val < h[i], change h[i] to val
        self.h[i] = val
        while (i != 0 and self.h[self.parent(i)] > self.h[i]):
            self.h[self.parent(i)], self.h[i] = self.h[i], self.h[self.parent(i)]
            i = self.parent(i)

    def get_min(self):
        return self.h[0] if self.size else None

    def delete_key(self, i):
        self.decrease_val(i, float('-inf'))
        self.extract_min()

    def push(self, val):
        i = self.size

        self.size += 1
        self.h.append(val)

        while (i != 0 and self.h[self.parent(i)] > self.h[i]):
            self.h[self.parent(i)], self.h[i] = self.h[i], self.h[self.parent(i)]
            i = self.parent(i)

    def min_heapify(self, i):
        ll = self.left(i)
        rr = self.right(i)
        small = i

        if ll < self.size and self.h[ll] < self.h[small]:
            small = ll

        if rr < self.size and self.h[rr] < self.h[small]:
            small = rr

        if small != i:
            self.h[i], self.h[small] = self.h[small], self.h[i]
            self.min_heapify(small)


class MinHeapq:
    """Implement heap by heapq"""
    def __init__(self):
        self.h = []

    @staticmethod
    def parent(i):
        return (i - 1) // 2

    def push(self, val):
        heappush(self.h, val)

    def get_min(self):
        return self.h[0] if self.h else None

    def extract_min(self):
        if self.h:
            return heappop(self.h)
        else:
            return None

    def decrease_val(self, i, val):
        # val < h[i], change h[i] to val
        self.h[i] = val
        while (i != 0 and self.h[self.parent(i)] > self.h[i]):
            self.h[self.parent(i)], self.h[i] = self.h[i], self.h[self.parent(i)]
            i = self.parent(i)

    def delete_key(self, i):
        self.decrease_val(i, float('-inf'))
        self.extract_min()


if __name__ == '__main__':
    mh = MinHeap() # Implement heap by list
    # mh = MinHeapq() # Implement heap by heapq
    mh.push(3)
    mh.push(2)
    mh.delete_key(1)
    print(mh.h)

    mh.push(15)
    mh.push(5)
    mh.push(4)
    mh.push(45)
    print(mh.h)
    print(mh.extract_min())
    print(mh.h)
    print(mh.get_min())

    mh.decrease_val(2, 1)
    print(mh.h)
    print(mh.get_min())
